File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/promises/extensions/specs/FuturePromise.java
package com.lasthopesoftware.bluewater.shared.promises.extensions.specs;

import android.support.annotation.NonNull;

import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.propagation.CancellationProxy;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class FuturePromise<Resolution> implements Future<Resolution> {

	private final CancellationProxy cancellationProxy = new CancellationProxy();
	private final Promise<Void> promise;

	private Resolution resolution;
	private Throwable rejection;
	private boolean isCompleted;

	public FuturePromise(Promise<Resolution> promise) {
		cancellationProxy.doCancel(promise);
		this.promise =
			promise
				.then(r -> {
					resolution = r;
					isCompleted = true;
					return null;
				}, e -> {
					rejection = e;
					isCompleted = true;
					return null;
				});
	}

	@Override
	public boolean cancel(boolean mayInterruptIfRunning) {
		if (isCompleted) return false;

		cancellationProxy.run();
		return true;
	}

	@Override
	public boolean isCancelled() {
		return cancellationProxy.isCancelled();
	}

	@Override
	public boolean isDone() {
		return isCompleted;
	}

	@Override
	public Resolution get() throws InterruptedException, ExecutionException {
		final CountDownLatch countDownLatch = new CountDownLatch(1);
		promise
			.then(r -> {
				countDownLatch.countDown();
				return null;
			});
		countDownLatch.await();

		return getResolution();
	}

	@Override
	public Resolution get(long timeout, @NonNull TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
		final CountDownLatch countDownLatch = new CountDownLatch(1);
		promise
			.then(r -> {
				countDownLatch.countDown();
				return null;
			});

		if (countDownLatch.await(timeout, unit))
			return getResolution();

		throw new TimeoutException();
	}

	private Resolution getResolution() throws ExecutionException {
		if (rejection == null)
			return resolution;

		throw new ExecutionException(rejection);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/shared/promises/extensions/FuturePromise.kt
package com.lasthopesoftware.bluewater.shared.promises.extensions

import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.handoff.promises.propagation.CancellationProxy
import java.util.concurrent.*

fun <Resolution> Promise<Resolution>.toFuture() = FuturePromise(this)

class FuturePromise<Resolution>(promise: Promise<Resolution>) : Future<Resolution?> {
	private val cancellationProxy = CancellationProxy()
	private val promise: Promise<Unit>
	private var resolution: Resolution? = null
	private var rejection: Throwable? = null
	private var isCompleted = false

	override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
		if (isCompleted) return false
		cancellationProxy.run()
		return true
	}

	override fun isCancelled(): Boolean {
		return cancellationProxy.isCancelled
	}

	override fun isDone(): Boolean {
		return isCompleted
	}

	@Throws(InterruptedException::class, ExecutionException::class)
	override fun get(): Resolution? {
		val countDownLatch = CountDownLatch(1)
		promise.then { countDownLatch.countDown() }
		countDownLatch.await()
		return getResolution()
	}

	@Throws(InterruptedException::class, ExecutionException::class, TimeoutException::class)
	override fun get(timeout: Long, unit: TimeUnit): Resolution? {
		val countDownLatch = CountDownLatch(1)
		promise.then {	countDownLatch.countDown() }
		if (countDownLatch.await(timeout, unit)) return getResolution()
		throw TimeoutException()
	}

	@Throws(ExecutionException::class)
	private fun getResolution(): Resolution? {
		rejection?.also { throw ExecutionException(rejection) }
		return resolution
	}

	init {
		cancellationProxy.doCancel(promise)
		this.promise = promise
			.then({ r: Resolution ->
				resolution = r
				isCompleted = true
			}, { e: Throwable? ->
				rejection = e
				isCompleted = true
			})
	}
}
